auslib/web/admin/views/base.py

```diff
@@ -69,17 +69,53 @@ def __init__(self, *args, **kwargs):
     @handleGeneralExceptions("POST")
     def post(self, *args, **kwargs):
         self.log.debug("processing POST request to %s" % request.path)
-        with dbo.begin() as trans:
-            return self._post(*args, transaction=trans, **kwargs)
+        trans = dbo.begin()
+        # Transactions are automatically rolled back by the context manager if
+        # _post raises an Exception, but we need to make sure they are also
+        # rolled back if the View returns any sort of error.
+        try:
+            ret = self._post(*args, transaction=trans, **kwargs)
+            if ret.status_code >= 400:
+                trans.rollback()
+            else:
+                trans.commit()
+            return ret
+        except:
+            trans.rollback()
+            raise
+        finally:
+            trans.close()
 
     @handleGeneralExceptions("PUT")
     def put(self, *args, **kwargs):
         self.log.debug("processing PUT request to %s" % request.path)
-        with dbo.begin() as trans:
-            return self._put(*args, transaction=trans, **kwargs)
+        trans = dbo.begin()
+        try:
+            ret = self._put(*args, transaction=trans, **kwargs)
+            if ret.status_code >= 400:
+                trans.rollback()
+            else:
+                trans.commit()
+            return ret
+        except:
+            trans.rollback()
+            raise
+        finally:
+            trans.close()
 
     @handleGeneralExceptions("DELETE")
     def delete(self, *args, **kwargs):
         self.log.debug("processing DELETE request to %s" % request.path)
-        with dbo.begin() as trans:
-            return self._delete(*args, transaction=trans, **kwargs)
+        trans = dbo.begin()
```

Comment:
Maybe the transaction behavior in POST, PUT and DELETE, can be in a decorator, something like this:

https://pastebin.mozilla.org/9069305